<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>compareCells compute differential gene expression between selected
territories for single cell type. — compareCells • vesalius</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="compareCells compute differential gene expression between selected
territories for single cell type. — compareCells"><meta property="og:description" content="compareCells compute differential gene expression between selected
territories for single cell type."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">vesalius</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/vesalius.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Vesalius_Analysis/Vesalius_analysis.html">Vesalius Manuscript Analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/patrickCNMartin/Vesalius/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>compareCells compute differential gene expression between selected
territories for single cell type.</h1>
    <small class="dont-index">Source: <a href="https://github.com/patrickCNMartin/Vesalius/blob/main/R/Territory_identity.R" class="external-link"><code>R/Territory_identity.R</code></a></small>
    <div class="hidden name"><code>compareCells.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>compareCells compute differential gene expression between selected
territories for single cell type.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">compareCells</span><span class="op">(</span></span>
<span>  <span class="va">image</span>,</span>
<span>  <span class="va">counts</span>,</span>
<span>  <span class="va">cells</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  query <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  logFC <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  pval <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  minPct <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  minBar <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  morphologyFactorSeed <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  morphologyFactorQuery <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>image</dt>
<dd><p>a dataframe containing territory information</p></dd>


<dt>counts</dt>
<dd><p>seurat object containing counts. Alternatively, matrix or
sparse matrix. Colnames as barcodes and rownames as genes.</p></dd>


<dt>cells</dt>
<dd><p>character vector containing barcodes associated with cell type
of interest.</p></dd>


<dt>seed</dt>
<dd><p>Integer or vector of integers describing territories to be
included in group 1 for differential gene expression analysis.</p></dd>


<dt>query</dt>
<dd><p>Integer or vector of integers describing territories to be
included in group 2 for differential gene expression analysis. Default = NULL</p></dd>


<dt>method</dt>
<dd><p>character describing the statistical test to use in order to
extract differantial gene expression (currently only wilcox and t.test)</p></dd>


<dt>logFC</dt>
<dd><p>numeric describing minimum log fold change value for
differential gene expression. Default set at 0.25.</p></dd>


<dt>pval</dt>
<dd><p>numeric for pval threshold. Default set at 0.05</p></dd>


<dt>minPct</dt>
<dd><p>numeric defining the minimum percentage of cells that should
contain any given gene.</p></dd>


<dt>minBar</dt>
<dd><p>integer defining minimum number of barcodes in a territory.</p></dd>


<dt>morphologyFactorSeed</dt>
<dd><p>Integer or vector of integers describing growth
or shrink rate of seed territories</p></dd>


<dt>morphologyFactorQuery</dt>
<dd><p>Integer or vector of integers describing growth
or shrink rate of query territories</p></dd>


<dt>verbose</dt>
<dd><p>logical - progress message output</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A data.frame (tibble) containing differential gene expression as well
p.value,
logFC,
seedPct (percentage of cells containing gene in first group),
queryPct (percentage of cells containing gene in second group),
seedTerritory (territory used as group 1)
queryTerritory (territory used as group 2)</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>compareCells provides a way to compare the expression of cells
between different territories. If you have a cell type annotations associated
to each of the barcodes present in your data, you can select your cell 
type of interest and parse the barcodes to the compareCell function.</p>
<p>Vesalius will internally subset your data and only compare
barcodes present in your territories of interest for differential 
gene expressioon.</p>
<p>The cells argument should be supplied as a vector of barcodes containing
all barcodes of the cell type of interest. Theoretically, you can supply
more than one cell type in this vector but Vesalius will treat them as single
cell type.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Patrick C.N. Martin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

