<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>compareClusters computes differential gene expression between seurat clusters — compareClusters • vesalius</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="compareClusters computes differential gene expression between seurat clusters — compareClusters" />
<meta property="og:description" content="compareClusters computes differential gene expression between seurat clusters" />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">vesalius</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/vesalius.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/patrickCNMartin/Vesalius/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>compareClusters computes differential gene expression between seurat clusters</h1>
    <small class="dont-index">Source: <a href='https://github.com/patrickCNMartin/Vesalius/blob/main/R/Territory_identity.R'><code>R/Territory_identity.R</code></a></small>
    <div class="hidden name"><code>compareClusters.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>compareClusters computes differential gene expression between seurat clusters</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>compareClusters</span><span class='op'>(</span>
  <span class='va'>counts</span>,
  <span class='va'>seedCluster</span>,
  <span class='va'>queryCluster</span>,
  seed <span class='op'>=</span> <span class='cn'>NULL</span>,
  query <span class='op'>=</span> <span class='cn'>NULL</span>,
  method <span class='op'>=</span> <span class='st'>"wilcox"</span>,
  logFC <span class='op'>=</span> <span class='fl'>0.25</span>,
  pval <span class='op'>=</span> <span class='fl'>0.05</span>,
  minPct <span class='op'>=</span> <span class='fl'>0.05</span>,
  minBar <span class='op'>=</span> <span class='fl'>10</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>counts</th>
      <td><p>seurat object containing ALL counts. Alternatively, matrix or
sparse matrix of all counts. Colnames as barcodes and rownames as genes.</p></td>
    </tr>
    <tr>
      <th>seedCluster</th>
      <td><p>Seurat Object of seed clusters</p></td>
    </tr>
    <tr>
      <th>queryCluster</th>
      <td><p>Seurat Object of query clusters</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>integer describing the Seurat cluster (in seed - group 1) to be
used for Differential expression analysis. Default NULL</p></td>
    </tr>
    <tr>
      <th>query</th>
      <td><p>integer describing the Seurat cluster (in query - group 2) to be
used for Differential expression analysis. Default NULL</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>character describing the statistical test to use in order to
extract differential gene expression (currently only wilcox and t.test)</p></td>
    </tr>
    <tr>
      <th>logFC</th>
      <td><p>numeric describing minimum log fold change value for
differential gene expression. Default set at 0.25.</p></td>
    </tr>
    <tr>
      <th>pval</th>
      <td><p>numeric for pval threshold. Default set at 0.05</p></td>
    </tr>
    <tr>
      <th>minPct</th>
      <td><p>numeric defining the minimum percentage of cells that should
contain any given gene.</p></td>
    </tr>
    <tr>
      <th>minBar</th>
      <td><p>integer defining minimum number of barcodes in a territory.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>logical - progress message output.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data.frame (tibble) containing differential gene expression as well
p.value,
logFC,
seedPct (percentage of cells containing gene in first group),
queryPct (percentage of cells containing gene in second group),
seedTerritory (territory used as group 1)
queryTerritory (territory used as group 2)</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Territory isolation provides a convenient way to investigate
the finer detail of anatomy and cellular spatial distribution. In some case,
two territories separated in 2D space may contain barcodes that cluster
together and are described by similar cell type labels. The
<code>compareClusters</code> functions compares clusters in different territories
between each other. The count values are the overall count values that have
been normalized, scaled and centered all together. This ensure that the
barcodes are comparable. The territory isolation and clustering serves as a
way to isolate a specific set of barcodes that will be used for differential
gene expression analysis.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"Vesalius"</span><span class='op'>)</span></span>
<span class='r-in'></span>
<span class='r-in'></span>
<span class='r-in'><span class='op'>}</span></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Patrick C.N. Martin.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


