<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Vesalius: Quick Start • vesalius</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Vesalius: Quick Start">
<meta property="og:description" content="vesalius">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">vesalius</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Vesalius.html">Vesalius: Quick Start</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Vesalius_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Vesalius: Quick Start</h1>
                        <h4 class="author">Patrick C.N. Martin</h4>
            
            <h4 class="date">2021-10-28</h4>
      
      
      <div class="hidden name"><code>Vesalius.Rmd</code></div>

    </div>

    
    
<div id="vesalius-quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#vesalius-quick-start" class="anchor"></a>Vesalius: Quick Start</h1>
<p>Do you like to live in the fast lane? Do you sigh every time you see a long vignette? Do you wish there was a TLDR on everything?</p>
<p>I understand.</p>
<p>Welcome to the Vesalius quick start!</p>
<p>In this vignette, you will find a <strong>very</strong> short guide on how to work with Vesalius. Vesalius provides an internal data set taken from real Spatial transcrtiptomic data. For an in depth view of Vesalius, please refer to Vesalius_Analysis.Rmd.</p>
<p>For all intents and purposes, this is a dummy data set and should only be used to get a feel for the Vesalius package. We strongly advise you to use real and complete data sets to fully understand the Vesalius package and its benefits.</p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Vesalius is a tool to perform high-resolution in silico anatomization and molecular characterization from ST data without requiring companion images.</p>
<p>This is achieved by converting ST data into images: PCA loading values are embedded into RGB colour channels by taking the sum of the absolute value of all loading values associated to each bead on the ST array.</p>
<p>Image analysis techniques are applied to the image and anatomical territories can subsequently be extracted.</p>
</div>
<div id="loading" class="section level1">
<h1 class="hasAnchor">
<a href="#loading" class="anchor"></a>Loading</h1>
<p>First, let’s load the package and the data. Data originates from slide-seqV2 data (availble at (Single Cell Portal)[<a href="https://singlecell.broadinstitute.org/single_cell/study/SCP948/robust-decomposition-of-cell-type-mixtures-in-spatial-transcriptomics#study-download" class="uri">https://singlecell.broadinstitute.org/single_cell/study/SCP948/robust-decomposition-of-cell-type-mixtures-in-spatial-transcriptomics#study-download</a>]).</p>
<p>For sake of simplicity, Vesalius takes a small subset of this data set and stores it in the form of a Seurat object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">vesalius</span><span class="op">)</span></code></pre></div>
<pre><code>## Loading required package: Seurat</code></pre>
<pre><code>## Attaching SeuratObject</code></pre>
<pre><code>## Loading required package: imager</code></pre>
<pre><code>## Loading required package: magrittr</code></pre>
<pre><code>## Registered S3 method overwritten by 'imager':
##   method      from
##   plot.imlist</code></pre>
<pre><code>## 
## Attaching package: 'imager'</code></pre>
<pre><code>## The following object is masked from 'package:magrittr':
## 
##     add</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     convolve, spectrum</code></pre>
<pre><code>## The following object is masked from 'package:graphics':
## 
##     frame</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     save.image</code></pre>
<pre><code>## Loading required package: dplyr</code></pre>
<pre><code>## 
## Attaching package: 'dplyr'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre><code>## Loading required package: tidyr</code></pre>
<pre><code>## 
## Attaching package: 'tidyr'</code></pre>
<pre><code>## The following object is masked from 'package:imager':
## 
##     fill</code></pre>
<pre><code>## The following object is masked from 'package:magrittr':
## 
##     extract</code></pre>
<pre><code>## Loading required package: ggplot2</code></pre>
<pre><code>## Warning: replacing previous import 'imager::fill' by 'tidyr::fill' when loading
## 'vesalius'</code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">vesalius</span>,package <span class="op">=</span> <span class="st">"vesalius"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "vesalius"</code></pre>
</div>
<div id="pre-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#pre-processing" class="anchor"></a>Pre-processing</h1>
<p>Second, we pre-process the data using Seurat</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># First create a copy of raw data for later use</span>
<span class="co"># When working on isolated territories, we want to run pre-processing</span>
<span class="co"># on raw counts not on already transformed counts.</span>
<span class="va">copy</span> <span class="op">&lt;-</span> <span class="va">vesalius</span>

<span class="co"># Preprocessing Spatial transcriptomic data</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">vesalius</span><span class="op">)</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">vesalius</span>, nfeatures <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">vesalius</span><span class="op">)</span></code></pre></div>
<pre><code>## Centering and scaling data matrix</code></pre>
</div>
<div id="rgb-to-pca" class="section level1">
<h1 class="hasAnchor">
<a href="#rgb-to-pca" class="anchor"></a>RGB to PCA</h1>
<p>Third, we convert a count matrix (stored in a Seurat object) to RGB coded coordinates. We will only produce one slice (i.e. PC1 to PC3)</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rgbPCA.html">rgbPCA</a></span><span class="op">(</span><span class="va">vesalius</span>,slices <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:13:38  Running Principle Component Analysis in 1 slices 
## 2021-10-28 10:13:38  Converting Loading Values to RGB in slice 1 
## 2021-10-28 10:13:44  Normalising RGB values 
## #------------------------------------------------------------------------#</code></pre>
</div>
<div id="building-an-image" class="section level1">
<h1 class="hasAnchor">
<a href="#building-an-image" class="anchor"></a>Building an Image</h1>
<p>Forth, we create an actual image by using a combination of voronoi diagrams and tile rasterisation.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildImageArray.html">buildImageArray</a></span><span class="op">(</span><span class="va">vesalius</span>, sliceID <span class="op">=</span> <span class="fl">1</span>, invert <span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:13:44  Checking Vesalius Input - Using slice  1 
## 2021-10-28 10:13:44  Class coersion to numeric 
## 2021-10-28 10:13:44  Inverting Colours 
## 2021-10-28 10:13:44  Filtering outlier beads 
## 2021-10-28 10:13:44  Generating Voronoi Tesselation 
## 2021-10-28 10:13:44  Rasterising Tiles 
## 2021-10-28 10:13:52  Filtering Triangles that exceed area threshold
## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's have a look shall we?</span>
<span class="va">Vesalius_image</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/imagePlot.html">imagePlot</a></span><span class="op">(</span><span class="va">vesalius</span>,as.cimg <span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span>


<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">Vesalius_image</span><span class="op">)</span></code></pre></div>
<p><img src="Vesalius_files/figure-html/buildImage-1.png" width="576"></p>
</div>
<div id="applying-image-analysis-techniques" class="section level1">
<h1 class="hasAnchor">
<a href="#applying-image-analysis-techniques" class="anchor"></a>Applying Image Analysis techniques</h1>
<p>Firth, we can apply image analysis techniques such as histogram equalisation, variance regularisation, smoothing and segmentation.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Histogram EQ</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/equalizeHistogram.html">equalizeHistogram</a></span><span class="op">(</span><span class="va">vesalius</span>,sleft <span class="op">=</span> <span class="fl">2.5</span>, sright<span class="op">=</span><span class="fl">2.5</span>,invert <span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:13:58  Inverting Colours</code></pre>
<pre><code>## Warning in as.cimg.data.frame(.): Guessing image dimensions from maximum
## coordinate values

## Warning in as.cimg.data.frame(.): Guessing image dimensions from maximum
## coordinate values</code></pre>
<pre><code>## 2021-10-28 10:14:02  Equalizing Histogram 
## 2021-10-28 10:14:03  Rebuilding Data Frame from image 
## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Variance regularisation</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/regulariseImage.html">regulariseImage</a></span><span class="op">(</span><span class="va">vesalius</span>, lambda <span class="op">=</span> <span class="fl">10</span>,
                            niter <span class="op">=</span> <span class="fl">200</span>, normalise<span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------#</code></pre>
<pre><code>## Warning in as.cimg.data.frame(.): Guessing image dimensions from maximum
## coordinate values</code></pre>
<pre><code>## 2021-10-28 10:14:05  Regularising Image 
## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Smoothing</span>
<span class="co"># !!!Optional!!! -</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/smoothArray.html">smoothArray</a></span><span class="op">(</span><span class="va">vesalius</span>,method <span class="op">=</span> <span class="st">"box"</span>, box <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in as.cimg.data.frame(select(image, c("x", "y", "cc", "value"))):
## Guessing image dimensions from maximum coordinate values</code></pre>
<pre><code>## 2021-10-28 10:14:18  Smoothing Iteration  1 </code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Segmentation</span>
<span class="co"># This function provides smoothing internally.</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/iterativeSegmentation.array.html">iterativeSegmentation.array</a></span><span class="op">(</span><span class="va">vesalius</span>,
                                        colDepth <span class="op">=</span> <span class="fl">6</span>,
                                        smoothIter <span class="op">=</span> <span class="fl">20</span>,
                                        method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"iso"</span>,<span class="st">"median"</span><span class="op">)</span>,
                                        sigma<span class="op">=</span><span class="fl">1.5</span>,
                                        box <span class="op">=</span> <span class="fl">10</span>,
                                        useCenter <span class="op">=</span> <span class="cn">T</span>,
                                        invert <span class="op">=</span><span class="cn">T</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------#</code></pre>
<pre><code>## Warning in as.cimg.data.frame(select(image, c("x", "y", "cc", "value"))):
## Guessing image dimensions from maximum coordinate values</code></pre>
<pre><code>## 2021-10-28 10:14:20  Inverting Colours</code></pre>
<pre><code>## Warning in as.cimg.data.frame(.): Guessing image dimensions from maximum
## coordinate values</code></pre>
<pre><code>## Warning in as.cimg.data.frame(imgCopy): Guessing image dimensions from maximum
## coordinate values</code></pre>
<pre><code>## 2021-10-28 10:14:24  Smoothing Iteration  1 
2021-10-28 10:14:27  Smoothing Iteration  2 
2021-10-28 10:14:29  Smoothing Iteration  3 
2021-10-28 10:14:32  Smoothing Iteration  4 
2021-10-28 10:14:34  Smoothing Iteration  5 
2021-10-28 10:14:37  Smoothing Iteration  6 
2021-10-28 10:14:39  Smoothing Iteration  7 
2021-10-28 10:14:42  Smoothing Iteration  8 
2021-10-28 10:14:44  Smoothing Iteration  9 
2021-10-28 10:14:47  Smoothing Iteration  10 
2021-10-28 10:14:49  Smoothing Iteration  11 
2021-10-28 10:14:51  Smoothing Iteration  12 
2021-10-28 10:14:54  Smoothing Iteration  13 
2021-10-28 10:14:56  Smoothing Iteration  14 
2021-10-28 10:14:59  Smoothing Iteration  15 
2021-10-28 10:15:01  Smoothing Iteration  16 
2021-10-28 10:15:03  Smoothing Iteration  17 
2021-10-28 10:15:06  Smoothing Iteration  18 
2021-10-28 10:15:08  Smoothing Iteration  19 
2021-10-28 10:15:10  Smoothing Iteration  20 
## 2021-10-28 10:15:14  Segmentation Iteration  1 
## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Isolating territories from colour segments</span>
<span class="va">vesalius</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isolateTerritories.array.html">isolateTerritories.array</a></span><span class="op">(</span><span class="va">vesalius</span>,
                                     captureRadius <span class="op">=</span> <span class="fl">0.1</span>,
                                     minBar <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:15:14  Pooling territory  1 
2021-10-28 10:15:14  Pooling territory  2 
2021-10-28 10:15:14  Pooling territory  3 
2021-10-28 10:15:15  Pooling territory  4 
2021-10-28 10:15:15  Pooling territory  5 
2021-10-28 10:15:16  Pooling territory  6 
## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">vesalius</span></code></pre></div>
<pre><code>## Registered S3 method overwritten by 'cli':
##   method     from         
##   print.boxx spatstat.geom</code></pre>
<pre><code>## # A tibble: 1,114,014 × 8
##    barcodes           x     y    cc value  tile cluster territory
##    &lt;chr&gt;          &lt;dbl&gt; &lt;dbl&gt; &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;chr&gt;    
##  1 AAGCAGTATGCGAT   146     1     1 0.865     0       3 1        
##  2 AACGCTTAGAATCT   146     1     1 0.865     0       3 1        
##  3 AGCTGTTGAGCTCC   146     1     1 0.865     0       3 1        
##  4 AACGCTTAGAATCT   146     2     1 0.865     0       3 1        
##  5 AACGCTTAGAATCT   146     3     1 0.865     0       3 1        
##  6 AAGCAGTATGCGAT   147     3     1 0.865     0       3 1        
##  7 AGCTGTTGAGCTCC   147     3     1 0.865     0       3 1        
##  8 GACGTATATCTGCA   516     3     1 0.523     0       4 2        
##  9 GATAATGGTCACCC   516     3     1 0.523     0       4 2        
## 10 AACGCTTAGAATCT   146     4     1 0.865     0       3 1        
## # … with 1,114,004 more rows</code></pre>
</div>
<div id="visualisation" class="section level1">
<h1 class="hasAnchor">
<a href="#visualisation" class="anchor"></a>Visualisation</h1>
<p>Sixth, we can visualise the territories.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Let's have a look!</span>
<span class="va">imgTerritory</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/territoryPlot.html">territoryPlot</a></span><span class="op">(</span><span class="va">vesalius</span>, randomise <span class="op">=</span> <span class="cn">TRUE</span>,cex <span class="op">=</span><span class="fl">15</span> , cex.pt<span class="op">=</span><span class="fl">3.5</span><span class="op">)</span>



<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">imgTerritory</span><span class="op">)</span></code></pre></div>
<p><img src="Vesalius_files/figure-html/viz-1.png" width="576"></p>
</div>
<div id="territory-extraction-and-markers" class="section level1">
<h1 class="hasAnchor">
<a href="#territory-extraction-and-markers" class="anchor"></a>Territory extraction and markers</h1>
<p>Finally, we can extract territories to run a more in depth analysis. This returns a Seurat object with all selected territories. We can also find differential gene expression between territories.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># extract terriotry from seurat object</span>
<span class="va">ter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractTerritories.html">extractTerritories</a></span><span class="op">(</span><span class="va">vesalius</span>,<span class="va">copy</span>,seedID<span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:15:19  Extracting territory 1 
## 2021-10-28 10:15:19  Converting to pixset and morphing territory</code></pre>
<pre><code>## Warning in as.cimg.data.frame(.): Guessing image dimensions from maximum
## coordinate values</code></pre>
<pre><code>## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ter</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 23264 features across 11 samples within 1 assay 
## Active assay: Spatial (23264 features, 0 variable features)</code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get all markers related to all territories</span>
<span class="va">all</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractAllMarkers.html">extractAllMarkers</a></span><span class="op">(</span><span class="va">vesalius</span>,<span class="va">copy</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:15:19  Checking and extracting Counts 
## ===&gt; 1 VS Remaining &lt;=== 
===&gt; 10 VS Remaining &lt;=== 
===&gt; 11 VS Remaining &lt;=== 
===&gt; 12 VS Remaining &lt;=== </code></pre>
<pre><code>## Warning: Territory 12 does not contain enough cells.
## 
##         Territory will be skipped</code></pre>
<pre><code>## ===&gt; 13 VS Remaining &lt;=== 
===&gt; 14 VS Remaining &lt;=== 
===&gt; 15 VS Remaining &lt;=== 
===&gt; 16 VS Remaining &lt;=== </code></pre>
<pre><code>## Warning: Territory 16 does not contain enough cells.
## 
##         Territory will be skipped</code></pre>
<pre><code>## ===&gt; 17 VS Remaining &lt;=== 
===&gt; 2 VS Remaining &lt;=== 
===&gt; 3 VS Remaining &lt;=== 
===&gt; 4 VS Remaining &lt;=== 
===&gt; 5 VS Remaining &lt;=== 
===&gt; 6 VS Remaining &lt;=== 
===&gt; 7 VS Remaining &lt;=== </code></pre>
<pre><code>## Warning: Territory 7 does not contain enough cells.
## 
##         Territory will be skipped</code></pre>
<pre><code>## ===&gt; 8 VS Remaining &lt;=== 
===&gt; 9 VS Remaining &lt;=== 
===&gt; isolated VS Remaining &lt;=== </code></pre>
<pre><code>## Warning: Territory isolated does not contain enough cells.
## 
##         Territory will be skipped</code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">all</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 8
##   genes p.value p.value.adj seedPct queryPct  logFC seedTerritory queryTerritory
##   &lt;chr&gt;   &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;         
## 1 Actb  1.82e-2   1           0.273   0.566  -0.802 1             Remaining     
## 2 Akt2  7.35e-6   0.00248     0.273   0.0308  0.342 1             Remaining     
## 3 Aldoa 2.96e-2   1           0.364   0.586  -0.815 1             Remaining     
## 4 Apod  1.50e-2   1           0.273   0.0809  0.400 1             Remaining     
## 5 Apoe  1.55e-3   0.518       0.909   0.623   0.565 1             Remaining     
## 6 Aqp4  1.62e-7   0.0000549   0.364   0.0405  0.330 1             Remaining</code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get markers associated with on or more territories</span>
<span class="va">one</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractMarkers.html">extractMarkers</a></span><span class="op">(</span><span class="va">vesalius</span>, <span class="va">copy</span>, seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:15:47  Checking and extracting Counts 
## 2021-10-28 10:15:47  Extracting Seed territories 
## 2021-10-28 10:15:47  Extracting Query territories 
## 2021-10-28 10:15:47  Computing Differentially Expressed Genes
## ===&gt; 1  VS Remaining &lt;=== 
## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">one</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 8
##   genes p.value p.value.adj seedPct queryPct  logFC seedTerritory queryTerritory
##   &lt;chr&gt;   &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;         
## 1 Actb  1.87e-2   1           0.273   0.564  -0.800 "1 "          Remaining     
## 2 Akt2  1.01e-5   0.00339     0.273   0.0317  0.341 "1 "          Remaining     
## 3 Aldoa 2.96e-2   1           0.364   0.586  -0.814 "1 "          Remaining     
## 4 Apod  1.48e-2   1           0.273   0.0807  0.400 "1 "          Remaining     
## 5 Apoe  1.58e-3   0.525       0.909   0.625   0.564 "1 "          Remaining     
## 6 Aqp4  1.55e-7   0.0000520   0.364   0.0403  0.331 "1 "          Remaining</code></pre>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># comparing territories</span>
<span class="va">two</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractMarkers.html">extractMarkers</a></span><span class="op">(</span><span class="va">vesalius</span>, <span class="va">copy</span>, seed <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## #------------------------------------------------------------------------# 
## 2021-10-28 10:15:48  Checking and extracting Counts 
## 2021-10-28 10:15:48  Extracting Seed territories 
## 2021-10-28 10:15:48  Extracting Query territories 
## 2021-10-28 10:15:48  Computing Differentially Expressed Genes
## ===&gt; 1  VS 2 3  &lt;=== 
## #------------------------------------------------------------------------#</code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">two</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 8
##   genes p.value p.value.adj seedPct queryPct  logFC seedTerritory queryTerritory
##   &lt;chr&gt;   &lt;dbl&gt;       &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;         &lt;chr&gt;         
## 1 Actb  1.87e-2   1           0.273   0.564  -0.800 "1 "          "2 3 "        
## 2 Akt2  1.01e-5   0.00339     0.273   0.0317  0.341 "1 "          "2 3 "        
## 3 Aldoa 2.96e-2   1           0.364   0.586  -0.814 "1 "          "2 3 "        
## 4 Apod  1.48e-2   1           0.273   0.0807  0.400 "1 "          "2 3 "        
## 5 Apoe  1.58e-3   0.525       0.909   0.625   0.564 "1 "          "2 3 "        
## 6 Aqp4  1.55e-7   0.0000520   0.364   0.0403  0.331 "1 "          "2 3 "</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Patrick C.N. Martin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
